diff -uprN "a/hw/acpi/aml-build.c" "b/hw/acpi/aml-build.c"
--- "a/hw/acpi/aml-build.c"	2026-05-20 23:52:20.423756983 +0800
+++ "b/hw/acpi/aml-build.c"	2026-05-20 23:52:20.427756983 +0800
@@ -1150,6 +1150,28 @@ Aml *aml_device(const char *name_format,
     return var;
 }
 
+Aml *aml_power_resource(const char *resource_name, uint8_t system_level,
+							uint16_t resource_order)
+{
+	Aml *var = aml_bundle(0x84  /* PowerResOp */, AML_EXT_PACKAGE);
+
+	build_append_namestring(var->buf, "%s", resource_name);
+	build_append_byte(var->buf, system_level);
+	build_append_byte(var->buf, resource_order & 0xFF);
+	build_append_byte(var->buf, resource_order >> 8);
+
+	return var;
+}
+
+Aml *aml_thermal_zone(const char *tz_name)
+{
+	Aml *var = aml_bundle(0x85 /* ThermalZoneOp */, AML_EXT_PACKAGE);
+
+	build_append_namestring(var->buf, "%s", tz_name) ;
+
+	return var;
+}
+
 /* ACPI 1.0b: 6.4.1 ASL Macros for Resource Descriptors */
 Aml *aml_resource_template(void)
 {
@@ -2263,7 +2285,7 @@ void build_fadt(GArray *tbl, BIOSLinker
     /* ACPI1.0: INT_MODEL, ACPI2.0+: Reserved */
     build_append_int_noprefix(tbl, f->int_model /* Multiple APIC */, 1);
     /* Preferred_PM_Profile */
-    build_append_int_noprefix(tbl, 0 /* Unspecified */, 1);
+    build_append_int_noprefix(tbl, 1 /* Desktop */, 1);
     build_append_int_noprefix(tbl, f->sci_int, 2); /* SCI_INT */
     build_append_int_noprefix(tbl, f->smi_cmd, 4); /* SMI_CMD */
     build_append_int_noprefix(tbl, f->acpi_enable_cmd, 1); /* ACPI_ENABLE */
diff -uprN "a/hw/i386/acpi-build.c" "b/hw/i386/acpi-build.c"
--- "a/hw/i386/acpi-build.c"	2026-05-20 23:52:20.415756983 +0800
+++ "b/hw/i386/acpi-build.c"	2026-05-20 23:52:20.419756983 +0800
@@ -845,6 +845,135 @@ static void build_acpi0017(Aml *table)
     aml_append(table, scope);
 }
 
+static void build_fan_device(Aml *table)
+{
+	Aml *sb_scope = aml_scope("\\_SB");
+	
+	Aml *power_res = aml_power_resource("PR01", 0, 0);
+	Aml *method = aml_method("_STA", 0, AML_SERIALIZED);
+	aml_append(method, aml_return(aml_int(0x0F)));
+	aml_append(power_res, method);
+
+	method=aml_method("_ON", 0, AML_SERIALIZED);
+	aml_append(power_res, method);
+	method=aml_method("_OFF", 0, AML_SERIALIZED);
+	aml_append(power_res, method);
+	
+	aml_append(sb_scope, power_res);
+	
+	Aml *dev = aml_device("FAN0");
+	aml_append(dev, aml_name_decl("_HID", aml_eisaid("PNP0C0B")));
+	aml_append(dev, aml_name_decl("_UID", aml_int(0)));
+
+	Aml *pkg = aml_package(1);
+	aml_append(pkg, aml_name("%s", "PR01"));
+	aml_append(dev, aml_name_decl("_PR0", pkg));
+
+    aml_append(sb_scope, dev);
+    aml_append(table, sb_scope);
+}
+
+static void build_thermal_zone(Aml *table)
+{
+	Aml *sb_scope = aml_scope("\\_SB");
+	Aml *tz = aml_thermal_zone("TZ00");
+
+	Aml *method = aml_method("_AC0", 0, AML_NOTSERIALIZED) ;
+	aml_append(method, aml_return(aml_int(3730)));
+	aml_append(tz, method);
+
+	Aml *pkg = aml_package(1);
+	aml_append(pkg, aml_name("%s", "\\_SB.FAN0"));
+	aml_append(tz, aml_name_decl("_AL0", pkg));
+
+	method = aml_method("_TMP", 0, AML_NOTSERIALIZED) ;
+	aml_append(method, aml_return(aml_int(3630)));
+	aml_append(tz, method) ;
+
+	method = aml_method("_CRT", 0, AML_NOTSERIALIZED) ;
+	aml_append(method, aml_return(aml_int(3730)));
+	aml_append(tz, method);
+
+	aml_append(sb_scope, tz);
+	aml_append(table, sb_scope);
+}
+
+static void build_PNP0000(Aml *table)
+{
+    Aml *dev, *scope;
+
+    scope =  aml_scope("_SB");
+    dev = aml_device("IPIC");
+    aml_append(dev, aml_name_decl("_HID", aml_eisaid("PNP0000")));
+
+    aml_append(dev, aml_name_decl("_STA", aml_int(0x0B)));
+
+    aml_append(scope, dev);
+    aml_append(table, scope);
+}
+
+static void build_PNP0C0C(Aml *table)
+{
+    Aml *dev, *scope, *method;
+
+    scope =  aml_scope("_SB");
+    dev = aml_device("PWRB");
+
+    aml_append(dev, aml_name_decl("_HID", aml_eisaid("PNP0C0C")));
+    aml_append(dev, aml_name_decl("PBST",  aml_string("One")));
+
+	method = aml_method("_STA", 0, AML_NOTSERIALIZED);
+	aml_append(method, aml_return(aml_int(0x0F)));
+    aml_append(dev, method);
+
+    aml_append(scope, dev);
+    aml_append(table, scope);
+}
+
+static void build_PNP0C0E(Aml *table)
+{
+    Aml *dev, *scope;
+
+    scope =  aml_scope("_SB");
+    dev = aml_device("SLPB");
+    aml_append(dev, aml_name_decl("_HID", aml_eisaid("PNP0C0E")));
+
+    aml_append(dev, aml_name_decl("_STA", aml_int(0x0B)));
+
+    aml_append(scope, dev);
+    aml_append(table, scope);
+}
+
+static void build_PNP0C14(Aml *table)
+{
+    Aml *dev, *scope;
+
+    scope =  aml_scope("_SB.PCI0");
+    dev = aml_device("WMI1");
+    aml_append(dev, aml_name_decl("_HID", aml_string("PNP0C14")));
+
+    aml_append(dev, aml_name_decl("_STA", aml_int(0x0B)));
+
+    aml_append(scope, dev);
+    aml_append(table, scope);
+}
+
+static void build_PNP0D80(Aml *table)
+{
+    Aml *dev, *scope;
+
+    scope =  aml_scope("_SB");
+    dev = aml_device("PEPD");
+    aml_append(dev, aml_name_decl("_HID", aml_string("INT33A1")));
+    aml_append(dev, aml_name_decl("_CID", aml_eisaid("PNP0D80")));
+
+
+    aml_append(dev, aml_name_decl("_STA", aml_int(0x0B)));
+
+    aml_append(scope, dev);
+    aml_append(table, scope);
+}
+
 static void
 build_dsdt(GArray *table_data, BIOSLinker *linker,
            AcpiPmInfo *pm, AcpiMiscInfo *misc,
@@ -938,6 +1067,15 @@ build_dsdt(GArray *table_data, BIOSLinke
         build_q35_pci0_int(dsdt);
     }
 
+	build_fan_device(dsdt);
+	build_thermal_zone(dsdt);
+	
+	build_PNP0000(dsdt);
+	build_PNP0C0C(dsdt);
+	build_PNP0C0E(dsdt);
+	build_PNP0C14(dsdt);
+	build_PNP0D80(dsdt);
+
     if (misc->has_hpet) {
         build_hpet_aml(dsdt);
     }
@@ -1630,6 +1768,8 @@ build_dmar_q35(GArray *table_data, BIOSL
  *
  * Helpful to speedup Windows guests and ignored by others.
  */
+
+/*
 static void
 build_waet(GArray *table_data, BIOSLinker *linker, const char *oem_id,
            const char *oem_table_id)
@@ -1638,13 +1778,39 @@ build_waet(GArray *table_data, BIOSLinke
                         .oem_table_id = oem_table_id };
 
     acpi_table_begin(&table, table_data);
-    /*
+    
      * Set "ACPI PM timer good" flag.
      *
      * Tells Windows guests that our ACPI PM timer is reliable in the
      * sense that guest can read it only once to obtain a reliable value.
      * Which avoids costly VMExits caused by guest re-reading it unnecessarily.
-     */
+     
+    build_append_int_noprefix(table_data, 1 << 1  ACPI PM timer good , 4);
+    acpi_table_end(linker, &table);
+}
+*/
+
+/*
+ * Intel ACPI Component Architecture
+ * AML/ASL+ Disassembler version 20251212 (64-bit version)
+ * Copyright (c) 2000 - 2025 Intel Corporation
+ * 
+ * Disassembly of /home/user/wsmt.dat
+ *
+ * ACPI Data Table [WSMT]
+ *
+ * Format: [HexOffset DecimalOffset ByteLength]  FieldName : FieldValue (in hex)
+ */
+
+static void
+build_wsmt(GArray *table_data, BIOSLinker *linker, const char *oem_id,
+           const char *oem_table_id)
+{
+    AcpiTable table = { .sig = "WSMT", .rev = 1, .oem_id = oem_id,
+                        .oem_table_id = oem_table_id };
+
+    acpi_table_begin(&table, table_data);
+
     build_append_int_noprefix(table_data, 1 << 1 /* ACPI PM timer good */, 4);
     acpi_table_end(linker, &table);
 }
@@ -2077,7 +2243,8 @@ void acpi_build(AcpiBuildTables *tables,
     }
 
     acpi_add_table(table_offsets, tables_blob);
-    build_waet(tables_blob, tables->linker, x86ms->oem_id, x86ms->oem_table_id);
+    //build_waet(tables_blob, tables->linker, x86ms->oem_id, x86ms->oem_table_id);
+    build_wsmt(tables_blob, tables->linker, x86ms->oem_id, x86ms->oem_table_id);
 
     /* Add tables supplied by user (if any) */
     for (u = acpi_table_first(); u; u = acpi_table_next(u)) {
diff -uprN "a/hw/i386/fw_cfg.c" "b/hw/i386/fw_cfg.c"
--- "a/hw/i386/fw_cfg.c"	2026-05-20 23:52:20.459756982 +0800
+++ "b/hw/i386/fw_cfg.c"	2026-05-20 23:52:20.459756982 +0800
@@ -225,7 +225,7 @@ void fw_cfg_add_acpi_dsdt(Aml *scope, FW
     uint8_t io_size = object_property_get_bool(obj, "dma_enabled", NULL) ?
         ROUND_UP(FW_CFG_CTL_SIZE, 4) + sizeof(dma_addr_t) :
         FW_CFG_CTL_SIZE;
-    Aml *dev = aml_device("FWCF");
+    Aml *dev = aml_device("FWC");
     Aml *crs = aml_resource_template();
 
     aml_append(dev, aml_name_decl("_HID", aml_string("ASUS0002")));
diff -uprN "a/hw/smbios/smbios.c" "b/hw/smbios/smbios.c"
--- "a/hw/smbios/smbios.c"	2026-05-20 23:52:20.443756983 +0800
+++ "b/hw/smbios/smbios.c"	2026-05-20 23:52:20.447756982 +0800
@@ -559,6 +559,7 @@ bool smbios_skip_table(uint8_t type, boo
 #define T16_BASE 0x1000
 #define T17_BASE 0x1100
 #define T19_BASE 0x1300
+#define T27_BASE 0x1700
 #define T32_BASE 0x2000
 #define T41_BASE 0x2900
 #define T127_BASE 0x7F00
@@ -941,6 +942,22 @@ static void smbios_build_type_19_table(u
     SMBIOS_BUILD_TABLE_POST;
 }
 
+static void smbios_build_type_27_table(void)
+{
+    const char *desc_str = "CPU Fan";
+
+    SMBIOS_BUILD_TABLE_PRE(27, T27_BASE, true);
+
+    t->temperature_probe_handle = cpu_to_le16(0xFFFF);
+    t->device_type_and_status = 0x65;
+    t->cooling_unitGroup = 0;
+    t->oem_defined = 0;
+    t->nominal_speed = cpu_to_le16(0x1C20);
+    SMBIOS_TABLE_SET_STR(27,description, desc_str);
+
+    SMBIOS_BUILD_TABLE_POST;
+}
+
 static void smbios_build_type_32_table(void)
 {
     SMBIOS_BUILD_TABLE_PRE(32, T32_BASE, true); /* required */
@@ -1161,6 +1178,8 @@ static bool smbios_get_tables_ep(Machine
                                    mem_array[i].length);
     }
 
+    smbios_build_type_27_table();
+
     /*
      * make sure 16 bit handle numbers in the headers of tables 19
      * and 32 do not overlap.
diff -uprN "a/include/hw/acpi/aml-build.h" "b/include/hw/acpi/aml-build.h"
--- "a/include/hw/acpi/aml-build.h"	2026-05-20 23:52:20.435756983 +0800
+++ "b/include/hw/acpi/aml-build.h"	2026-05-20 23:52:20.439756983 +0800
@@ -388,6 +388,9 @@ Aml *aml_error_device(void);
 /* Block AML object primitives */
 Aml *aml_scope(const char *name_format, ...) G_GNUC_PRINTF(1, 2);
 Aml *aml_device(const char *name_format, ...) G_GNUC_PRINTF(1, 2);
+Aml *aml_power_resource(const char *resource_name, uint8_t system_level,
+							uint16_t resource_order);
+Aml *aml_thermal_zone(const char *tz_name);
 Aml *aml_method(const char *name, int arg_count, AmlSerializeFlag sflag);
 Aml *aml_if(Aml *predicate);
 Aml *aml_else(void);
diff -uprN "a/include/hw/firmware/smbios.h" "b/include/hw/firmware/smbios.h"
--- "a/include/hw/firmware/smbios.h"	2026-05-20 23:52:20.451756982 +0800
+++ "b/include/hw/firmware/smbios.h"	2026-05-20 23:52:20.455756982 +0800
@@ -303,6 +303,17 @@ struct smbios_type_19 {
     uint64_t extended_ending_address;
 } QEMU_PACKED;
 
+/* SMBIOS type 27 – Cooling Device */
+struct smbios_type_27 {
+    struct smbios_structure_header header;
+    uint16_t temperature_probe_handle;
+    uint8_t device_type_and_status;
+    uint8_t cooling_unitGroup;
+    uint32_t oem_defined;
+    uint16_t nominal_speed;
+    uint8_t description ;
+} QEMU_PACKED;
+
 /* SMBIOS type 32 - System Boot Information */
 struct smbios_type_32 {
     struct smbios_structure_header header;
